### ISSUE TITLE
Dice_Dense input inconsistent with remaining segmentation loss functions

### ISSUE THREAD

**Issue by zsdqui** [NifTK/NiftyNet#339]
The dice_dense and dice_dense_nosquare functions in the loss._segmentation.py file seem inconsistent with the remainder of the loss functions in that they assume a one_hot encoded ground_truth. The remainder of the loss functions use categorical designations for the ground_truth input and then convert to one-hot encoding using the labels_to_one_hot function. 

For instance using dice_dense or dice_dense_square with the highres3dnet throws tensor dimensions error, which is fixed if the loss function is modified to one_hot encode its input

Sample config file throwing an error:
```
[NETWORK]
name = highres3dnet_large
activation_function = relu
batch_size = 4
decay = 0
reg_type = L2

# volume level preprocessing
volume_padding_size = 0
#histogram normalisation
normalisation = True
histogram_ref_file = /home/chingon/niftynet/histogram_4DLungNiftiOAR.txt
norm_type = percentile
cutoff = (0.01, 0.99)
whitening = False
normalise_foreground_only=False
foreground_type = otsu_plus
multimod_foreground_type = or
#window_sampling = balanced

[TRAINING]
optimiser = adam
sample_per_volume = 64
rotation_angle = (-5.0, 5.0)
scaling_percentage = (-5.0, 5.0)
# random_flipping_axes= 1
lr = 0.01
loss_type = niftynet.layer.loss_segmentation.dice_dense
starting_iter = 0
save_every_n = 100
max_iter = 500
max_checkpoints = 20
validation_every_n = 5
exclude_fraction_for_validation = 0.05
```

**Comment by Zach-ER** [comment]
Yes, this is true. 'Dense' refers to the fact that the labels are given as densely labelled, multi-channel images whereas they are normally given as integer labels. This functionality is useful if, for example, classes are not exclusive (for example, hierarchical labels).
